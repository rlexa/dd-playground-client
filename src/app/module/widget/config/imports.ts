import { CommonModule } from '@angular/common';
import { MatCardModule, MatListModule } from '@angular/material';
import { RouterModule } from '@angular/router';

export const imports = [CommonModule, RouterModule, MatCardModule, MatListModule];
