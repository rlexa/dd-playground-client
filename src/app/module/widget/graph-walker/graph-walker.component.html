<ng-template #tmplNode let-item let-header="header">
  <div ddFlexbox="vyt" class="contenthv childhv hovered-elevated">
    <h4 *ngIf="header" class="text-center">{{ header }}</h4>
    <div ddFlexbox="hyc">
      <label class="childh">{{ item.inc.length }}</label>
      <label ddFlexitem="1 0" class="childh text-center">{{ 'inc .. out' }}</label>
      <label class="childh">{{ item.out.length }}</label>
    </div>
    <app-simple-view [data]="item.data" [isDense]="true" [isWidthConstrained]="false"></app-simple-view>
  </div>
</ng-template>

<ng-template #tmplLink let-item let-header="header">
  <div ddFlexbox="vyt" class="contenthv childhv hovered-elevated">
    <h4 *ngIf="header" class="text-center">{{ header }}</h4>
    <app-simple-view [data]="item.data" [isDense]="true" [isWidthConstrained]="false"></app-simple-view>
  </div>
</ng-template>

<!-- START HERE -->

<mat-card>
  <mat-card-content>
    <div ddFlexbox="vyt">
      <div ddFlexbox="hysw">
        <div class="contenthv childhv hovered-elevated">
          <h2>DB Stats: Overall</h2>
          <app-simple-view [data]="dbState$ | async" [isDense]="true" [isWidthConstrained]="false"></app-simple-view>
        </div>

        <div class="contenthv childhv hovered-elevated">
          <h2>DB Stats: Node Types</h2>
          <app-simple-view [data]="dbNodeTypeCount$ | async" [isDense]="true" [isWidthConstrained]="false"></app-simple-view>
        </div>
      </div>

      <div ddFlexbox="hycw" class="contenthv childhv hovered-elevated">
        <label>Query first node of:</label>

        <mat-form-field class="margin-left childhv" hintLabel="Type">
          <mat-select [value]="curType$ | async" (selectionChange)="setCurType($event.value)">
            <mat-option *ngFor="let ii of dbTypes$ | async" [value]="ii"> {{ ii }} </mat-option>
          </mat-select>
        </mat-form-field>

        <mat-form-field class="margin-left childhv" hintLabel="Tag">
          <mat-select [value]="curTag$ | async" (selectionChange)="setCurTag($event.value)">
            <mat-option *ngFor="let ii of curTypeKeys$ | async" [value]="ii"> {{ ii }} </mat-option>
          </mat-select>
        </mat-form-field>

        <button *ngFor="let ii of CMPS" mat-mini-fab (click)="tryQuery(ii)">{{ ii }}</button>

        <mat-form-field class="margin-left childhv" hintLabel="Value">
          <input
            #valInput
            matInput
            [value]="curVal$ | async"
            (change)="setCurVal($event.target.value)"
            [matAutocomplete]="valAutoComplete"
            (focus)="setCurVal(valInput.value)"
            (keyup)="setCurVal(valInput.value)"
            (keyup.enter)="tryQuery()"
          />
          <mat-autocomplete #valAutoComplete="matAutocomplete" (optionSelected)="setCurVal($event.option.value)">
            <mat-option *ngFor="let ii of curTypeKeyTagValuesFiltered$ | async" [value]="ii">{{ ii }}</mat-option>
          </mat-autocomplete>
        </mat-form-field>
      </div>

      <div *ngIf="curNode$ | async as node" [ngStyle]="{display: 'grid', 'grid-template-areas': GRID, 'grid-template-columns': GRID_SIZE}">
        <div [ngStyle]="{'grid-area': GRID_FROM}">
          <div ddFlexbox="vye">
            <div *ngFor="let link of node.inc" ddFlexbox="hys" class="childh">
              <ng-container *ngTemplateOutlet="tmplNode; context: {$implicit: link.from}"></ng-container>
              <div ddFlexbox="hyc">
                <ng-container *ngTemplateOutlet="tmplLink; context: {$implicit: link}"></ng-container>
                <button mat-mini-fab (click)="setCurNode(link.from)">
                  <mat-icon>{{ 'go_left' | icon }}</mat-icon>
                </button>
              </div>
            </div>
          </div>
        </div>

        <div [ngStyle]="{'grid-area': GRID_CURRENT}">
          <div ddFlexbox="vyc" class="childh">
            <ng-container *ngTemplateOutlet="tmplNode; context: {$implicit: node, header: 'CURRENT'}"></ng-container>
          </div>
        </div>

        <div [ngStyle]="{'grid-area': GRID_TO}">
          <div ddFlexbox="vys" class="childh">
            <div *ngFor="let link of node.out" ddFlexbox="hys">
              <div ddFlexbox="hyc">
                <button mat-mini-fab (click)="setCurNode(link.to)">
                  <mat-icon>{{ 'go_right' | icon }}</mat-icon>
                </button>
                <ng-container *ngTemplateOutlet="tmplLink; context: {$implicit: link}"></ng-container>
              </div>
              <ng-container *ngTemplateOutlet="tmplNode; context: {$implicit: link.to}"></ng-container>
            </div>
          </div>
        </div>
      </div>
    </div>
  </mat-card-content>
</mat-card>
