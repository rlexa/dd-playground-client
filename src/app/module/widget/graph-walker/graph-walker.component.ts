import {ChangeDetectionStrategy, Component, OnDestroy, OnInit} from '@angular/core';
import {GraphskyService, IGraphskyNode} from 'src/app/module/service/graphsky-api';
import {RxCleanup, rxNext_} from 'dd-rxjs';
import {BehaviorSubject, combineLatest, of} from 'rxjs';
import {debounceTime, filter, map, withLatestFrom} from 'rxjs/operators';

export const TAG_TYPE = '_type';

const CMP_EQ = '=';
const CMP_NE = '!=';
const CMP_LE = '<=';
const CMP_LT = '<';
const CMP_GE = '>=';
const CMP_GT = '>';
const compareEqual = (cmp: string, aa: any, bb: any) => {
  let ret = false;
  switch (cmp) {
    case CMP_GT:
      ret = aa > bb;
      break;
    case CMP_GE:
      ret = aa >= bb;
      break;
    case CMP_LT:
      ret = aa < bb;
      break;
    case CMP_LE:
      ret = aa <= bb;
      break;
    case CMP_NE:
      ret = aa !== bb;
      break;
    case CMP_EQ:
    default:
      ret = aa === bb;
  }
  return ret;
};

@Component({
  selector: 'app-graph-walker',
  templateUrl: './graph-walker.component.html',
  changeDetection: ChangeDetectionStrategy.OnPush,
})
export class GraphWalkerComponent implements OnDestroy, OnInit {
  constructor(private readonly graphsky: GraphskyService) {}

  readonly GRID_FROM = 'from';
  readonly GRID_CURRENT = 'current';
  readonly GRID_TO = 'to';
  readonly GRID = [[this.GRID_FROM, this.GRID_CURRENT, this.GRID_TO]].map(line => '"' + line.join(' ') + '"').join(' ');
  readonly GRID_SIZE = '1fr auto 1fr';

  readonly CMPS = [CMP_LT, CMP_LE, CMP_NE, CMP_EQ, CMP_GE, CMP_GT];

  @RxCleanup() readonly curType$ = new BehaviorSubject<string>(null);
  @RxCleanup() readonly curTag$ = new BehaviorSubject<string>(null);
  @RxCleanup() readonly curVal$ = new BehaviorSubject<string>(null);
  @RxCleanup() readonly curNode$ = new BehaviorSubject<IGraphskyNode>(null);

  readonly dbState$ = this.graphsky.log$;
  readonly dbNodeTypeCount$ = this.graphsky.change$.pipe(
    debounceTime(200),
    map(() =>
      this.graphsky.query((nodes, _) =>
        nodes.reduce<Record<string, number>>(
          (acc, ii) => ({...acc, [ii.data[TAG_TYPE] as string]: (acc[ii.data[TAG_TYPE] as string] || 0) + 1}),
          {},
        ),
      ),
    ),
  );

  readonly dbTypes$ = this.dbNodeTypeCount$.pipe(map(val => Object.keys(val).sort()));
  readonly curTypeFirstNode$ = this.curType$.pipe(
    filter(type => !!type && type.length > 0),
    map(type => this.graphsky.query((nodes, _) => nodes.find(ii => ii.data[TAG_TYPE] === type))),
  );
  readonly curTypeKeys$ = this.curTypeFirstNode$.pipe(map(node => (node ? Object.keys(node.data).filter(ii => ii !== TAG_TYPE) : [])));
  readonly curTypeKeyTagValues$ = combineLatest([this.curType$, this.curTypeKeys$, this.curTag$]).pipe(
    map(([type, keys, tag]) =>
      !type || !keys || !keys.includes(tag)
        ? []
        : this.graphsky.query((nodes, _) =>
            nodes
              .filter(ii => ii.data[TAG_TYPE] === type && tag in ii.data && ii.data[tag] !== null && ii.data[tag] !== undefined)
              .map(ii => ii.data[tag].toString())
              .reduce<string[]>((acc, ii) => (acc.includes(ii) ? acc : [...acc, ii]), [])
              .sort(),
          ),
    ),
  );
  readonly curTypeKeyTagValuesFiltered$ = combineLatest([this.curVal$, this.curTypeKeyTagValues$]).pipe(
    map(([val, vals]) =>
      !val || !val.length ? vals || [] : (vals || []).filter(ii => ii.toLocaleLowerCase().includes(val.toLocaleLowerCase())).sort(),
    ),
  );

  setCurNode = rxNext_(this.curNode$);
  setCurType = rxNext_(this.curType$);
  setCurTag = rxNext_(this.curTag$);
  setCurVal = rxNext_(this.curVal$);

  ngOnDestroy() {}

  ngOnInit() {
    this.dbTypes$
      .pipe(
        withLatestFrom(this.curType$),
        filter(([vals, cur]) => vals.length > 0 && (!cur || !vals.includes(cur))),
        map(([vals]) => vals[0]),
      )
      .subscribe(this.setCurType);

    this.curTypeKeys$
      .pipe(
        withLatestFrom(this.curTag$),
        filter(([vals, cur]) => vals.length > 0 && (!cur || !vals.includes(cur))),
        map(([vals]) => vals[0]),
      )
      .subscribe(this.setCurTag);
  }

  tryQuery = (cmp = CMP_EQ) =>
    of([this.curType$.value, this.curTag$.value, this.curVal$.value])
      .pipe(
        filter(params => params.every(ii => !!ii && ii.length > 0)),
        map(([type, tag, val]) =>
          this.graphsky.query((nodes, _) => nodes.filter(ii => ii.data[TAG_TYPE] === type && compareEqual(cmp, ii.data[tag], val))),
        ),
        map(nodes => (nodes.length ? nodes[0] : null)),
      )
      .subscribe(this.setCurNode);
}
