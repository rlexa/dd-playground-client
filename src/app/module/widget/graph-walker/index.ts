export * from './graph-walker.component';
export * from './graph-walker.module';
