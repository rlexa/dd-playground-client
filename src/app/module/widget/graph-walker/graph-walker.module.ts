import {CommonModule} from '@angular/common';
import {NgModule} from '@angular/core';
import {MatAutocompleteModule} from '@angular/material/autocomplete';
import {MatButtonModule} from '@angular/material/button';
import {MatCardModule} from '@angular/material/card';
import {MatFormFieldModule} from '@angular/material/form-field';
import {MatIconModule} from '@angular/material/icon';
import {MatInputModule} from '@angular/material/input';
import {MatSelectModule} from '@angular/material/select';
import {RouterModule, Routes} from '@angular/router';
import {ROUTE_ROOT, ROUTE_WILDCARD} from 'src/app/routing';
import {FlexboxModule} from '../../directive/flexbox';
import {IconPipeModule} from '../../pipe/icon';
import {SimpleViewModule} from '../simple-view';
import {GraphWalkerComponent, TAG_TYPE} from './graph-walker.component';

const ROUTING: Routes = [
  {path: ROUTE_ROOT, component: GraphWalkerComponent, pathMatch: 'full'},
  {path: ROUTE_WILDCARD, redirectTo: ROUTE_ROOT},
];

@NgModule({
  imports: [
    CommonModule,
    MatAutocompleteModule,
    MatButtonModule,
    MatCardModule,
    MatFormFieldModule,
    MatIconModule,
    MatInputModule,
    MatSelectModule,
    FlexboxModule,
    IconPipeModule,
    SimpleViewModule,
    RouterModule.forChild(ROUTING),
  ],
  exports: [GraphWalkerComponent],
  declarations: [GraphWalkerComponent],
})
class GraphWalkerModule {}

export {GraphWalkerModule, GraphWalkerComponent, TAG_TYPE};
