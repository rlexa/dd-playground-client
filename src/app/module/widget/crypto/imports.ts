import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import {
  MatButtonModule,
  MatCardModule,
  MatFormFieldModule,
  MatInputModule,
  MatListModule,
} from '@angular/material';
import { RouterModule } from '@angular/router';
import { CryptoApiModule } from 'app/module/service/crypto-api';
import { SimpleViewModule } from 'app/module/widget/simple-view';
import { FlexboxModule } from 'app/module/directive/flexbox';

export const imports = [
  CommonModule,
  FormsModule,
  RouterModule,
  MatButtonModule,
  MatCardModule,
  MatFormFieldModule,
  MatInputModule,
  MatListModule,
  FlexboxModule,
  CryptoApiModule,
  SimpleViewModule,
];
