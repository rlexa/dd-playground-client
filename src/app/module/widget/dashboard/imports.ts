import { CommonModule } from '@angular/common';
import { MatButtonModule, MatToolbarModule, MatTooltipModule } from '@angular/material';
import { RouterModule } from '@angular/router';
import { IconPipeModule } from 'app/module/pipe/icon';
import { RoutedContentModule } from 'app/module/widget/routed-content';
import { VersionModule } from 'app/module/widget/version';
import { WidgetsModule } from 'app/widgets/widgets.module';
import { FlexboxModule } from 'dd-flexbox';
import 'hammerjs';

export const imports = [
  CommonModule, RouterModule,
  MatButtonModule, MatToolbarModule, MatTooltipModule, FlexboxModule,
  IconPipeModule, VersionModule,
  // remove
  RoutedContentModule, WidgetsModule,
];
