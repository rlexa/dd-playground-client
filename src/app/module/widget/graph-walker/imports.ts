import { CommonModule } from '@angular/common';
import { MatAutocompleteModule, MatButtonModule, MatCardModule, MatFormFieldModule, MatIconModule, MatInputModule, MatSelectModule } from '@angular/material';
import { RouterModule } from '@angular/router';
import { IconPipeModule } from 'app/module/pipe/icon';
import { GraphskyApiModule } from 'app/module/service/graphsky-api';
import { SimpleViewModule } from 'app/module/widget/simple-view';
import { FlexboxModule } from 'dd-flexbox';

export const imports = [
  CommonModule, RouterModule,
  MatAutocompleteModule, MatButtonModule, MatCardModule, MatFormFieldModule, MatIconModule, MatInputModule, MatSelectModule, FlexboxModule,
  GraphskyApiModule, IconPipeModule, SimpleViewModule,
];
