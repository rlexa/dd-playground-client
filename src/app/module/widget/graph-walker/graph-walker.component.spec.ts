import {async, TestBed} from '@angular/core/testing';
import {NoopAnimationsModule} from '@angular/platform-browser/animations';
import {RouterTestingModule} from '@angular/router/testing';
import {GraphskyService} from 'src/app/module/service/graphsky-api';
import {GraphWalkerComponent} from './graph-walker.component';
import {imports} from './imports';

describe('GraphWalkerComponent', () => {
  beforeEach(async(() => {
    TestBed.configureTestingModule({
      imports: [NoopAnimationsModule, RouterTestingModule, ...imports],
      declarations: [GraphWalkerComponent],
      providers: [GraphskyService],
    }).compileComponents();
  }));

  test('is created', () => {
    const fixture = TestBed.createComponent(GraphWalkerComponent);
    fixture.detectChanges();
    expect(fixture).toMatchSnapshot();
  });
});
