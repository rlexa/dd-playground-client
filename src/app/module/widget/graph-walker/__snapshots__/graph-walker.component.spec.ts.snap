// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GraphWalkerComponent is created 1`] = `
<app-graph-walker
  CMPS={[Function Array]}
  GRID={[Function String]}
  GRID_CURRENT={[Function String]}
  GRID_FROM={[Function String]}
  GRID_SIZE={[Function String]}
  GRID_TO={[Function String]}
  __ngContext__={[Function LRootView]}
  curNode$={[Function BehaviorSubject]}
  curTag$={[Function BehaviorSubject]}
  curType$={[Function BehaviorSubject]}
  curTypeFirstNode$={[Function AnonymousSubject]}
  curTypeKeyTagValues$={[Function Observable]}
  curTypeKeyTagValuesFiltered$={[Function Observable]}
  curTypeKeys$={[Function AnonymousSubject]}
  curVal$={[Function BehaviorSubject]}
  dbNodeTypeCount$={[Function Observable]}
  dbState$={[Function BehaviorSubject]}
  dbTypes$={[Function Observable]}
  graphsky={[Function GraphskyService]}
  setCurNode={[Function Function]}
  setCurTag={[Function Function]}
  setCurType={[Function Function]}
  setCurVal={[Function Function]}
  tryQuery={[Function Function]}
>
  <mat-card>
    <mat-card-content>
      <div
        ddflexbox="vyt"
        ng-reflect-dd-flexbox="vyt"
      >
        <div
          ddflexbox="hysw"
          ng-reflect-dd-flexbox="hysw"
        >
          <div
            class="contenthv childhv hovered-elevated"
          >
            <h2>
              DB Stats: Overall
            </h2>
            <app-simple-view
              ng-reflect-data="[object Object]"
              ng-reflect-is-dense="true"
              ng-reflect-is-width-constrained="false"
            />
          </div>
          <div
            class="contenthv childhv hovered-elevated"
          >
            <h2>
              DB Stats: Node Types
            </h2>
            <app-simple-view
              ng-reflect-is-dense="true"
              ng-reflect-is-width-constrained="false"
            />
          </div>
        </div>
        <div
          class="contenthv childhv hovered-elevated"
          ddflexbox="hycw"
          ng-reflect-dd-flexbox="hycw"
        >
          <label>
            Query first node of:
          </label>
          <mat-form-field
            class="margin-left childhv"
            hintlabel="Type"
            ng-reflect-hint-label="Type"
          >
            <mat-select>
            </mat-select>
          </mat-form-field>
          <mat-form-field
            class="margin-left childhv"
            hintlabel="Tag"
            ng-reflect-hint-label="Tag"
          >
            <mat-select>
            </mat-select>
          </mat-form-field>
          <button
            mat-mini-fab=""
          >
            &lt;
          </button>
          <button
            mat-mini-fab=""
          >
            &lt;=
          </button>
          <button
            mat-mini-fab=""
          >
            !=
          </button>
          <button
            mat-mini-fab=""
          >
            =
          </button>
          <button
            mat-mini-fab=""
          >
            &gt;=
          </button>
          <button
            mat-mini-fab=""
          >
            &gt;
          </button>
          <mat-form-field
            class="margin-left childhv"
            hintlabel="Value"
            ng-reflect-hint-label="Value"
          >
            <input
              matinput=""
              ng-reflect-autocomplete="[object Object]"
            />
            <mat-autocomplete>
            </mat-autocomplete>
          </mat-form-field>
        </div>
      </div>
    </mat-card-content>
  </mat-card>
</app-graph-walker>
`;
